### ISSUE TITLE
Add autocast support

### ISSUE THREAD

**Issue by johnbradley** [Imageomics/pybioclip#108]
Add [torch.autocast](https://pytorch.org/docs/stable/amp.html#automatic-mixed-precision-package-torch-amp) support potentially allowing faster performance.

Thanks to @samuelstevens for suggesting this.

**Comment by johnbradley** [comment]
# TreeOfLifeClassifier
When testing this I encountered an issue where loading the tree of life text embeddings were float32 even inside the autocast context. To work around this I had to cast them to correct data type.

```
import torch
from bioclip import TreeOfLifeClassifier, Rank
device = "mps"
with torch.autocast(device_type=device):
    classifier = TreeOfLifeClassifier(device=device)
    classifier.txt_embeddings = classifier.txt_embeddings.to(torch.get_autocast_dtype(device))
    result = classifier.predict("examples/images/Ursus-arctos.jpeg", Rank.SPECIES, k=1)
result
```

Just like CustomLabelsClassifier the image features, text features, and logits were float16 and probabilities were float32

**Comment by johnbradley** [comment]
To implement this my thought is to add a boolean `autocast` setting:
- `bioclip predict --device cuda --autocast Ursus-arctos.jpeg`
- `CustomLabelsClassifier(autocast=True, ...)`
- `TreeOfLifeClassifier(autocast=True, ...)`

When true `autocast` would cause the text embeddings loaded to be cast to the appropriate dtype and the predict method would use `torch.amp.autocast()`.

**Comment by samuelstevens** [comment]
Oh it might be that autocast is only supported on CUDA devices, I'm not sure if mps supports 16-bit operations. I basically only optimize for CUDA operations in my research.

**Comment by johnbradley** [comment]
In my testing on mps autocast seemed to work(data types where float16). I'll do some benchmarking to see if actually improves performance on MPS. Sometimes mps switches back to cpu.

**Comment by johnbradley** [comment]
## MPS
Testing on my Mac with autocast resulted in **20% slower** predictions.

## CPU
Testing on my Mac CPU with autocast was **5x slower**! I also tested on OSC with CPU and found similar slowness.


## CUDA
Testing on OSC with autocast was **5% faster**.


---
My Mac is a Apple M2 Pro Laptop.

**Comment by hlapp** [comment]
The performance improvement on CUDA doesn't look very convincing, and also not convincing enough to justify making the code more complex?

**Comment by samuelstevens** [comment]
Interesting. Not sure why that's the case, I almost always experience a speedup. It might be that for a batch size of 1, the casting operations are too expensive to make up for the faster matmul operations.

**Comment by samuelstevens** [comment]
yeah probably converting PIL images into tensors + moving them from CPU to GPU. In training pipelines with high throughput we end up using multiple processes for dataloading and pre-emptively moving them on the GPU to try and avoid these bottlenecks. But since BioCLIP is pretty small (~120M params?) the GPU cost of matmuls is likely pretty small. Worth a try; thanks for benchmarking it!

**Comment by johnbradley** [comment]
Closing this since the improvement was minor on CUDA and not a major bottleneck ([preprocessing was](https://github.com/Imageomics/pybioclip/blob/35ad338bb904d18b295dabf8b23377a343273c06/src/bioclip/predict.py#L166-L175)).
